server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -134,6 +162,238 @@ public List<MetricAggregatorInfo> generateMetricAggregatorInfos(MapperService ma
+    }
+
+    /**
+     * Builds the star tree using sorted and aggregated star-tree Documents
+     *
+     * @param starTreeDocumentIterator   contains the sorted and aggregated documents
+     * @param fieldNumberAcrossStarTrees maintains a counter for the number of star-tree fields
+     * @param starTreeDocValuesConsumer  consumes the generated star-tree docValues
+     * @throws IOException when we are unable to build star-tree
+     */
+    public void build(
+        Iterator<StarTreeDocument> starTreeDocumentIterator,
+        AtomicInteger fieldNumberAcrossStarTrees,
+        DocValuesConsumer starTreeDocValuesConsumer
+    ) throws IOException {
+        int numSegmentStarTreeDocument = totalSegmentDocs;
+
+        while (starTreeDocumentIterator.hasNext()) {
+            appendToStarTree(starTreeDocumentIterator.next());
+        }
+        int numStarTreeDocument = numStarTreeDocs;
+        logger.debug("Generated star tree docs : [{}] from segment docs : [{}]", numStarTreeDocument, numSegmentStarTreeDocument);
+
+        if (numStarTreeDocs == 0) {
+            // serialize the star tree data
+            serializeStarTree(numSegmentStarTreeDocument);
+            return;
+        }
+
+        constructStarTree(rootNode, 0, numStarTreeDocs);
+        int numStarTreeDocumentUnderStarNode = numStarTreeDocs - numStarTreeDocument;
+        logger.debug(
+            "Finished constructing star-tree, got [ {} ] tree nodes and [ {} ] starTreeDocument under star-node",
+            numStarTreeNodes,
+            numStarTreeDocumentUnderStarNode
+        );
+
+        createAggregatedDocs(rootNode);
+        int numAggregatedStarTreeDocument = numStarTreeDocs - numStarTreeDocument - numStarTreeDocumentUnderStarNode;
+        logger.debug("Finished creating aggregated documents : {}", numAggregatedStarTreeDocument);
+
+        // Create doc values indices in disk
+        createSortedDocValuesIndices(starTreeDocValuesConsumer, fieldNumberAcrossStarTrees);
+
+        // serialize star-tree
+        serializeStarTree(numSegmentStarTreeDocument);
+    }
+
+    private void serializeStarTree(int numSegmentStarTreeDocument) throws IOException {
+        // serialize the star tree data
+        long dataFilePointer = dataOut.getFilePointer();
+        long totalStarTreeDataLength = StarTreeUtils.serializeStarTree(dataOut, rootNode, numStarTreeNodes);
+
+        // serialize the star tree meta
+        StarTreeUtils.serializeStarTreeMetadata(
+            metaOut,
+            starTreeField,
+            writeState,
+            metricAggregatorInfos,
+            numSegmentStarTreeDocument,
+            dataFilePointer,
+            totalStarTreeDataLength
+        );
+    }
+
+    private void createSortedDocValuesIndices(DocValuesConsumer docValuesConsumer, AtomicInteger fieldNumberAcrossStarTrees)
+        throws IOException {
+        List<SortedNumericDocValuesWriterHelper> dimensionWriters = new ArrayList<>();
```

Comment:
Do we need to differentiate as long as we form unique field names for both dimensions and metrics ?

Reply:
yes, it allows us to have custom behaviour for the metrics, where we need to check the aggregated value type before we can perform add value.
Plus, the distinction makes it more readable.